core/focus_manager.ts

```diff
@@ -0,0 +1,305 @@
+/**
+ * @license
+ * Copyright 2025 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import type {IFocusableNode} from './interfaces/i_focusable_node.js';
+import type {IFocusableTree} from './interfaces/i_focusable_tree.js';
+import * as dom from './utils/dom.js';
+
+/**
+ * Type declaration for returning focus to FocusManager upon completing an
+ * ephemeral UI flow (such as a dialog).
+ *
+ * See FocusManager.takeEphemeralFocus for more details.
+ */
+export type ReturnEphemeralFocus = () => void;
+
+/**
+ * A per-page singleton that manages Blockly focus across one or more
+ * IFocusableTrees, and bidirectionally synchronizes this focus with the DOM.
+ *
+ * Callers that wish to explicitly change input focus for select Blockly
+ * components on the page should use the focus functions in this manager.
+ *
+ * The manager is responsible for handling focus events from the DOM (which may
+ * may arise from users clicking on page elements) and ensuring that
+ * corresponding IFocusableNodes are clearly marked as actively/passively
+ * highlighted in the same way that this would be represented with calls to
+ * focusNode().
+ */
+export class FocusManager {
+  focusedNode: IFocusableNode | null = null;
+  registeredTrees: Array<IFocusableTree> = [];
+
+  private currentlyHoldsEphemeralFocus: boolean = false;
+
+  constructor(
+    addGlobalEventListener: (type: string, listener: EventListener) => void,
+  ) {
+    // Register root document focus listeners for tracking when focus leaves all
+    // tracked focusable trees.
+    addGlobalEventListener('focusin', (event) => {
+      if (!(event instanceof FocusEvent)) return;
+
+      // The target that now has focus.
+      const activeElement = document.activeElement;
+      let newNode: IFocusableNode | null = null;
+      if (
+        activeElement instanceof HTMLElement ||
+        activeElement instanceof SVGElement
+      ) {
+        // If the target losing focus maps to any tree, then it should be
+        // updated. Per the contract of findFocusableNodeFor only one tree
+        // should claim the element.
+        const matchingNodes = this.registeredTrees.map((tree) =>
+          tree.findFocusableNodeFor(activeElement),
+        );
+        newNode = matchingNodes.find((node) => !!node) ?? null;
```

Comment:
nit, but I feel like there is something kind of funny about using the nullish coalescing operator to make the value null instead of undefined. Why not just use undefined? You could change line 48 to `let newNode: IFocusableNode | undefined;` and then remove this bit

Reply:
I agree that it does seem odd. I suspect there's some argument here about using an intentional absent (null) vs. an unintentional absent (undefined). I'll think on this more, but I'm happy to defer to your suggestion if I can't come up with a strong argument.

Reply:
Looks like it basically has to be `null | undefined` so I went ahead with that. Couldn't find strong advice on the best practices of one vs. the other, unfortunately.